File: src/Java/tree/103.二叉树的锯齿形层序遍历.java
/*
 * @lc app=leetcode.cn id=103 lang=java
 *
 * [103] 二叉树的锯齿形层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (root == null) return res;
        Deque<TreeNode> q = new ArrayDeque<TreeNode>();
        q.offer(root);
        boolean isLeft = true;
        while (!q.isEmpty()) {
            Deque<Integer> level = new ArrayDeque<Integer>();
            int currentLength = q.size();
            for (int i = 0; i < currentLength; i++) {
                TreeNode node = q.poll();
                if (isLeft) {
                    level.offerLast(node.val);
                } else {
                    level.offerFirst(node.val);
                }
                if (node.left != null) q.offer(node.left);
                if (node.right != null) q.offer(node.right);
            }
            res.add(new ArrayList<Integer>(level));
            isLeft = !isLeft;
        }
        return res;
    }
}
// @lc code=end

File: src/Javascript/tree/103.二叉树的锯齿形层序遍历.js
/*
 * @lc app=leetcode.cn id=103 lang=javascript
 *
 * [103] 二叉树的锯齿形层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    const res = [];
    if (root === null) return res;
    const q = [];
    let isLeft = true;
    q.push(root);
    while (q.length) {
        const currentLength = q.length;
        res.push([]);
        for (let i = 0; i < currentLength; i++) {
            const node = q.shift();
            if (isLeft) {
                res[res.length - 1].push(node.val);
            } else {
                res[res.length - 1].unshift(node.val);
            }
            if (node.left) q.push(node.left);
            if (node.right) q.push(node.right);
        }
        isLeft = !isLeft;
    }
    return res;
};
// @lc code=end

